middle_end/flambda2/from_lambda/lambda_to_lambda_transforms.ml

```diff
@@ -110,7 +110,422 @@ let rec_catch_for_for_loop env loc ident start stop
   in
   env, lam
 
-let transform_primitive env (prim : L.primitive) args loc =
+type initialize_array_element_width =
+  | Thirty_two of { zero_init : L.lambda }
+  | Sixty_four_or_more
+
+let initialize_array0 env loc ~length array_set_kind width ~(init : L.lambda)
+    creation_expr =
+  let array = Ident.create_local "array" in
+  (* If the element size is 32-bit, zero-initialize the last 64-bit word, to
+     ensure reproducibility. *)
+  (* CR mshinwell: why does e.g. caml_make_unboxed_int32_vect not do this? *)
+  let maybe_zero_init_last_field =
+    match width with
+    | Sixty_four_or_more -> L.lambda_unit
+    | Thirty_two { zero_init } ->
+      let zero_init_last_field =
+        L.Lprim
+          ( Parraysetu (array_set_kind, Ptagged_int_index),
+            (* [Popaque] is used to conceal the out-of-bounds write. *)
+            [Lprim (Popaque L.layout_unit, [Lvar array], loc); length; zero_init],
+            loc )
+      in
+      let length_is_greater_than_zero_and_is_one_mod_two =
+        L.Lprim
+          ( Psequand,
+            [ Lprim (Pintcomp Cgt, [Lconst (L.const_int 0); length], loc);
+              Lprim
+                ( Pintcomp Cne,
+                  [ Lprim (Pmodint Unsafe, [length; Lconst (L.const_int 2)], loc);
+                    Lconst (L.const_int 0) ],
+                  loc ) ],
+            loc )
+      in
+      L.Lifthenelse
+        ( length_is_greater_than_zero_and_is_one_mod_two,
+          zero_init_last_field,
+          L.lambda_unit,
+          L.layout_unit )
+  in
+  let env, initialize =
+    let index = Ident.create_local "index" in
+    rec_catch_for_for_loop env loc index
+      (Lconst (L.const_int 0))
+      (L.Lprim (Psubint, [length; Lconst (L.const_int 1)], loc))
+      Upto
+      (Lprim
+         ( Parraysetu (array_set_kind, Ptagged_int_index),
+           [Lvar array; Lvar index; init],
+           loc ))
+  in
+  let term =
+    L.Llet
+      ( Strict,
+        Pvalue { raw_kind = Pgenval; nullable = Non_nullable },
+        array,
+        creation_expr,
+        Lsequence
+          (maybe_zero_init_last_field, Lsequence (initialize, Lvar array)) )
+  in
+  env, Transformed term
+
+let initialize_array env loc ~length array_set_kind width ~init creation_expr =
+  match init with
+  | None -> env, Transformed creation_expr
+  | Some init ->
+    initialize_array0 env loc ~length array_set_kind width ~init creation_expr
+
+let makearray_dynamic_singleton name (mode : L.locality_mode) ~length ~init loc
```

Comment:
@mshinwell Still working on tests, but found this along the way. The following, when compiled with `ocamlopt`:
```ocaml
external[@layout_poly] make_vect : ('a : any_non_null) . int -> 'a -> 'a array =
  "%makearray_dynamic"

let _ = make_vect 100 #3.14
```

Gives this error:
```
:(.text+0x1f): undefined reference to `caml_make_unboxed_float64_vect'
collect2: error: ld returned 1 exit status
```

Reply:
Thanks, fixed.